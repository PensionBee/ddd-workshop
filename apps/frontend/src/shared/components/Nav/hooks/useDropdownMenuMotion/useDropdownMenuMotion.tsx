import { useEffect, useState } from "react";

type useDropdownMenuMotionProps = {
  defaultOpen?: boolean;
  delay?: number;
};

export const useDropdownMenuMotion = (props?: useDropdownMenuMotionProps) => {
  const { defaultOpen = false, delay = 200 } = props || {};
  const [isOpen, setIsOpen] = useState(defaultOpen || false);
  const [isVisible, setIsVisible] = useState(false);
  const [motion, setMotion] = useState<"from-end" | "to-end" | null>(null);

  const toggleOpen = () => setIsOpen((state) => !state);

  useEffect(() => {
    if (isOpen) {
      setIsVisible(true);
      setMotion("from-end");
    } else {
      if (isVisible) {
        setMotion("to-end");
        const timeout = setTimeout(() => setIsVisible(false), delay); // Wait for animation to finish
        return () => clearTimeout(timeout);
      }
    }
  }, [isVisible, isOpen, delay]);

  return { isVisible, motion, toggleOpen, setIsOpen };
};
