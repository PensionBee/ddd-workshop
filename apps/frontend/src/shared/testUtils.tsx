import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import type { Breakpoints } from "./styling/breakpoints";
import { BREAKPOINTS } from "./styling/breakpoints";
import { theme } from "./styling/theme";
import { vi } from "vitest"

type ReactTestingLibraryProviderProps = {
  children: React.ReactNode;
};

// eslint-disable-next-line react-refresh/only-export-components
const ReactTestingLibraryProviders: React.FC<
  ReactTestingLibraryProviderProps
> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const customRender = (ui: React.ReactElement, options?: any) => {
  return render(ui, { wrapper: ReactTestingLibraryProviders, ...options });
};

// re-export everything
// eslint-disable-next-line react-refresh/only-export-components
export * from "@testing-library/react";

// override render method
export { customRender as render };

export const sleep = (delay: number) => {
  return new Promise((resolve) => setTimeout(resolve, delay));
};

export const mockMediaMatches = (breakpoint: Breakpoints) => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: (query: string) => {
      const breakpointValue = BREAKPOINTS[breakpoint];
      const queryValue = parseInt(query.replace(/^\D+/g, ""));
      const breakpointValues = [...Object.values(BREAKPOINTS)];
      if (!breakpointValues.includes(queryValue)) {
        throw new Error(
          `Invalid breakpoint value: ${queryValue}, this likely means there has been a change to the way that breakpoints are setup.`
        );
      }
      return {
        matches: queryValue <= breakpointValue,
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
      };
    },
  });
};

export const routerRender = (ui: React.ReactNode, pathname = "/") => {
  return customRender(
    <MemoryRouter initialEntries={[{ pathname }]}>{ui}</MemoryRouter>
  );
};
