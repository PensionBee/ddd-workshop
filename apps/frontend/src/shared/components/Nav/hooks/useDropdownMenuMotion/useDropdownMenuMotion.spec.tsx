import { act, fireEvent, render, sleep } from "@shared/testUtils";
import { useDropdownMenuMotion } from "./useDropdownMenuMotion";

describe("useDropdownMenuMotion", () => {
  it("should return default values", () => {
    const VISIBLE_VALUE = "visibleValue";
    const MOTION_VALUE = "motionValue";
    const Component = () => {
      const { isVisible, motion } = useDropdownMenuMotion();
      return (
        <>
          <span data-value={`${isVisible}`}>{VISIBLE_VALUE}</span>
          <span data-value={`${motion}`}>{MOTION_VALUE}</span>
        </>
      );
    };
    const { getByText, asFragment } = render(<Component />);
    const visibleElement = getByText(VISIBLE_VALUE);
    const motionElement = getByText(MOTION_VALUE);
    expect(visibleElement).toHaveAttribute("data-value", "false");
    expect(motionElement).toHaveAttribute("data-value", "null");
    expect(asFragment()).toMatchSnapshot();
  });

  it("should toggle isVisible", async () => {
    const VISIBLE_VALUE = "visibleValue";
    const TOGGLE_BUTTON = "toggleButton";
    const Component = () => {
      const { isVisible, toggleOpen } = useDropdownMenuMotion();
      return (
        <>
          <span data-value={`${isVisible}`}>{VISIBLE_VALUE}</span>
          <button onClick={toggleOpen}>{TOGGLE_BUTTON}</button>
        </>
      );
    };
    const { getByText, asFragment } = render(<Component />);
    const visibleElement = getByText(VISIBLE_VALUE);
    const toggleButton = getByText(TOGGLE_BUTTON);
    expect(visibleElement).toHaveAttribute("data-value", "false");
    act(() => fireEvent.click(toggleButton));
    await act(() => sleep(500)); // Longer than default delay
    expect(visibleElement).toHaveAttribute("data-value", "true");
    act(() => fireEvent.click(toggleButton));
    await act(() => sleep(500)); // Longer than default delay
    expect(visibleElement).toHaveAttribute("data-value", "false");
    expect(asFragment()).toMatchSnapshot();
  });

  it("should take delay prop", async () => {
    const VISIBLE_VALUE = "visibleValue";
    const TOGGLE_BUTTON = "toggleButton";
    const Component = () => {
      const { isVisible, toggleOpen } = useDropdownMenuMotion({
        delay: 0,
      });
      return (
        <>
          <span data-value={`${isVisible}`}>{VISIBLE_VALUE}</span>
          <button onClick={toggleOpen}>{TOGGLE_BUTTON}</button>
        </>
      );
    };
    const { getByText, asFragment } = render(<Component />);
    const visibleElement = getByText(VISIBLE_VALUE);
    const toggleButton = getByText(TOGGLE_BUTTON);
    expect(visibleElement).toHaveAttribute("data-value", "false");
    act(() => fireEvent.click(toggleButton));
    await act(() => sleep(50)); // Shorter than default delay
    expect(visibleElement).toHaveAttribute("data-value", "true");
    act(() => fireEvent.click(toggleButton));
    await act(() => sleep(50)); // Shorter than default delay
    expect(visibleElement).toHaveAttribute("data-value", "false");
    expect(asFragment()).toMatchSnapshot();
  });

  it("should take defaultOpen prop", () => {
    const VISIBLE_VALUE = "visibleValue";
    const Component = () => {
      const { isVisible } = useDropdownMenuMotion({
        defaultOpen: true,
      });
      return (
        <>
          <span data-value={`${isVisible}`}>{VISIBLE_VALUE}</span>
        </>
      );
    };
    const { getByText, asFragment } = render(<Component />);
    const visibleElement = getByText(VISIBLE_VALUE);
    expect(visibleElement).toHaveAttribute("data-value", "true");
    expect(asFragment()).toMatchSnapshot();
  });
});
