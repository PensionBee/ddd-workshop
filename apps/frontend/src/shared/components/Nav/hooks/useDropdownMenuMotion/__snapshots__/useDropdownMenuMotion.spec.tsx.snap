// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useDropdownMenuMotion > should return default values 1`] = `
<DocumentFragment>
  <span
    data-value="false"
  >
    visibleValue
  </span>
  <span
    data-value="null"
  >
    motionValue
  </span>
</DocumentFragment>
`;

exports[`useDropdownMenuMotion > should take defaultOpen prop 1`] = `
<DocumentFragment>
  <span
    data-value="true"
  >
    visibleValue
  </span>
</DocumentFragment>
`;

exports[`useDropdownMenuMotion > should take delay prop 1`] = `
<DocumentFragment>
  <span
    data-value="false"
  >
    visibleValue
  </span>
  <button>
    toggleButton
  </button>
</DocumentFragment>
`;

exports[`useDropdownMenuMotion > should toggle isVisible 1`] = `
<DocumentFragment>
  <span
    data-value="false"
  >
    visibleValue
  </span>
  <button>
    toggleButton
  </button>
</DocumentFragment>
`;

exports[`useDropdownMenuMotion should return default values 1`] = `
<DocumentFragment>
  <span
    data-value="false"
  >
    visibleValue
  </span>
  <span
    data-value="null"
  >
    motionValue
  </span>
</DocumentFragment>
`;

exports[`useDropdownMenuMotion should take defaultOpen prop 1`] = `
<DocumentFragment>
  <span
    data-value="true"
  >
    visibleValue
  </span>
</DocumentFragment>
`;

exports[`useDropdownMenuMotion should take delay prop 1`] = `
<DocumentFragment>
  <span
    data-value="false"
  >
    visibleValue
  </span>
  <button>
    toggleButton
  </button>
</DocumentFragment>
`;

exports[`useDropdownMenuMotion should toggle isVisible 1`] = `
<DocumentFragment>
  <span
    data-value="false"
  >
    visibleValue
  </span>
  <button>
    toggleButton
  </button>
</DocumentFragment>
`;
